plan/relations.go

```diff
@@ -194,6 +232,24 @@ func (v *VirtualTableReadRel) ToProtoPlanRel() *proto.PlanRel {
 	}
 }
 
+func (v *VirtualTableReadRel) Copy(newInputs ...Rel) (Rel, error) {
+	return v, nil
+}
+
+func (v *VirtualTableReadRel) CopyWithExpressionRewrite(rewriteFunc func(expr.Expression) (expr.Expression, error), newInputs ...Rel) (Rel, error) {
+	newExprs, err := v.baseReadRel.copyExpressions(rewriteFunc)
+	if err != nil {
+		return nil, err
+	}
+	if newExprs == nil {
+		return nil, nil
+	}
+	nt := *v
+	nt.filter = newExprs[0]
+	nt.bestEffortFilter = newExprs[1]
+	return &nt, nil
```

Comment:
Rather than repeating this over and over, it feels like we could easily write a simple generic helper function, something like:

```go
func (b *baseReadRel) updateFilters(filter, bestEffortFilter expr.Expression) {
    b.filter, b.bestEffortFilter = filter, bestEffortFilter
}

type exprRewriter interface {
    copyExpressions(func(expr.Expression) (expr.Expression, error)) ([]expr.Expression, error)
    updateFilters(filter, bestEffortFilter expr.Expression)
}

func copyRewriteHelper[T exprRewriter](r *T, rewriteFunc func(expr.Expression) (expr.Expression, error)) (*T, error) {
    newExprs, err := r.copyExpressions(rewriteFunc)
    if err != nil {
        return nil, err
    }

    if newExprs == nil {
        return r, nil
    }

    nt := *r
    nt.updateFilters(newExprs[0], newExprs[1])
    return &nt, nil
}

...

func (v *VirtualTableReadRel) CopyWithExpressionRewrite(rewriteFunc func(expr.Expression) (expr.Expression, error), newInputs ...Rel) (Rel, error) {
    return copyRewriteHelper(v, rewriteFunc)
}
```

Reply:
I hit some issues when I tried this.
1. `func copyRewriteHelper[T exprRewriter](r *T, rewriteFunc RewriteFunc) (*T, error)`
       - r is a pointer to interface, complains about `unresolved copyReference`
       - r is a pointer to interface, so `nt := *r` won't copy the object?
2. `func copyRewriteHelper[T exprRewriter](r T, rewriteFunc RewriteFunc) (*T, error)`
       - `nt := *r` this throws error `invalid indirect of type T`. Looks like we cannot copy object using interface type variable.

Reply:
Below works, but still some code duplication of is there.

```
type exprRewriter interface {
	copyExpressions(rewriteFunc RewriteFunc) ([]expr.Expression, error)
	getExpressions() []expr.Expression
	updateFilters(filter []expr.Expression)
	copy() Rel
}

func copyRewriteHelper(n exprRewriter, rewriteFunc RewriteFunc) (Rel, error) {
	newExprs, err := n.copyExpressions(rewriteFunc)
	if err != nil {
		return nil, err
	}
	if slices.Equal(newExprs, n.getExpressions()) {
		return n.(Rel), nil
	}
	nt := n.copy()
	nt.(exprRewriter).updateFilters(newExprs)
	return nt, nil
}

func (n *NamedTableReadRel) copy() Rel {
	ntr := *n
	return &ntr
}

func (n *NamedTableReadRel) CopyWithExpressionRewrite1(rewriteFunc RewriteFunc, _ ...Rel) (Rel, error) {
	return copyRewriteHelper(n, rewriteFunc)
}
```